### DISCUSSION TITLE
Optional doesn't work well with null

### DISCUSSION THREAD

**Discussion by francoism90** [spatie/laravel-data#161] [Ideas]
Payload (request):
```
{
    name: 'Test',
    content: null,
}
```

Data object:
```
public function __construct(
        public string $name,
        public Optional|string $content,
    ) {
    }
```
Results in invalid data error. Is this intended? A workaround is to use `public Optional|string|null $content` or to filter null values out of the payload.

**Comment by juse-less** [comment]
I took a quick peek on this. It's definitely possible that I've missed the `Optional` reference somewhere.
That aside - There's a [`RuleInferrer` for `Sometimes`](https://github.com/spatie/laravel-data/blob/main/src/RuleInferrers/SometimesRuleInferrer.php#L13-L15), which checks if the property is `Optional`, and, if so, sets Laravels '[sometimes](https://laravel.com/docs/9.x/validation#validating-when-present)' rule.
From my understanding of 'sometimes', it'll skip validation if the whole attribute is missing, but use the other rules to validate it, if the attribute is present.
So it sounds like you'd need to add `null` as a type for it.

Depending on what you'd like to achieve, you could possibly use the Exclude rules to remove the attribute from the data, if the value is null.
Maybe something like
```php
public static function rules(array $payload): array
{
    return [
        'content' => [Rule::excludeIf(! isset($payload['content']))],
    ];
}
```

as outlined in the [Laravel documentation](https://laravel.com/docs/9.x/validation#rule-exclude-if).
It could be possible to use something like `Rule::excludeUnless()`, as well, but it wasn't directly written in the documentation and I don't have my editor available to check right now.

Also of note, you _may_ need to also add the `Sometimes` rule in that `Data::rules()` array, since we're overwriting rules.
Drawing blanks if one needs to or not.

Edit: So I don't forget - you may also set `Optional` yourself when you instantiate the class doing something like
```php
$data['content'] ??= Optional::create();

Data::from($data);
```
But I'm not sure that'd actually help with the validation.
([From the Data documentation](https://spatie.be/docs/laravel-data/v2/as-a-data-transfer-object/optional-properties) - scroll to the bottom.)

**Reply by toni-abrx** [reply]
Replying in hope it will be helpful to someone in the future.
The `Optional|string` type hint denotes that the property **may be omitted** but if it is present, it **must** be a `string`. That's why the null check fails. In case `null` values should be allowed, you need to type hint it like this:
```php
public function __construct(
        public string $name,
        public Optional|string|null $content, // add null type hint here
    ) {
    }
```
